File: samples/leetcode/arrays/contains_duplicates.rs
use std::collections::HashMap;
use std::error::Error;

struct Solution;

impl Solution {

    fn using_hash_map(nums: Vec<i32>) -> bool {
        let mut map = HashMap::new();
        for &num in nums.iter() {
            if map.contains_key(&num) {
                return true;
            }
            map.insert(num, 1);
        }
        return false;
    }
    pub fn contains_duplicate(nums: Vec<i32>) -> bool {
        return Self::using_hash_map(nums);
    }
}
fn main() -> Result<(), Box<dyn Error>> {
    let nums = vec![1,2,3,1];
    println!("{}", Solution::contains_duplicate(nums));
    let nums = vec![1,1,1,3,3,4,3,2,4,2];
    println!("{}", Solution::contains_duplicate(nums));
    let nums = vec![1,2,3,4];
    println!("{}", Solution::contains_duplicate(nums));

    Ok(())
}

File: bazel/samples/leetcode/arrays/contains_duplicates.cpp
#include <vector>
#include <iostream>
#include <unordered_map>

using namespace std;

class Solution {
public:

    bool usingHashMap(vector<int>& nums) {
        std::unordered_map<int, int> unorderedMap;
        for (auto &x : nums) {
            if (unorderedMap.contains(x)) {
                return true;
            }
            unorderedMap[x] = 1;
        }
        return false;
    }

    bool containsDuplicate(vector<int>& nums) {
        return usingHashMap(nums);
    }
};

int main() {
    //vector<int> nums = {1,2,3,1};
    //vector<int> nums = {1,1,1,3,3,4,3,2,4,2};
    vector<int> nums = {1,2,3,4};
    auto s = Solution();
    auto ret = s.containsDuplicate(nums);
    std::cout << std::boolalpha <<  "Returned " << ret << std::endl;
    return 0;
}